packages/kiota_abstractions/lib/src/enum_registry.dart

```diff
@@ -0,0 +1,63 @@
+part of '../kiota_abstractions.dart';
+
+class _EnumRegistration {
+  _EnumRegistration(this.enumType, this.cases);
+
+  final Type enumType;
+  final List<Enum> cases;
+}
+
+/// This is a helper class to register [Enum] cases so they can be used in
+/// serialization and deserialization.
+class EnumRegistry {
+  EnumRegistry._();
+
+  static final Map<Type, _EnumRegistration> _registrations = {};
```

Comment:
I think the solution is elegant, short of using annotations/attributes/an equivalent and reflection.
Or short of co-generating parsing and serialization methods (like we do in go)

My main concerns:
- Can I assume that to enums with the same name but in different namespaces won't collide? (I don't know enough about dart's type system)
- Can I assume this is thread safe for both registration but also when reading?

Reply:
Thanks for reviewing!

> Can I assume that to enums with the same name but in different namespaces won't collide?

I just tested this with a minimal project and it seems to work. AFAIK types in dart are internally represented by their full package name plus their type name (for example: "package:dart_kiota/dart_kiota.dart:HttpMethod" or similar).

That means enums from different packages with the same type name should be able to coexist in this map.

> Can I assume this is thread safe for both registration but also when reading?

Dart has a userland variant of "threads" they call "Isolate". IIRC these Isolates map 1:1 to OS threads and have their own memory area. This registry is _not_ safe to be used across different isolates. For that purpose, we'd need to find a different solution (there exists a package `shared_map` that could achieve this).

Reply:
Thanks for the context. So with the current implementation, it's possible that a consumer starts a request on thread 1 (UI, where they instantiated the client), defers it to a thread pool, and when the response is being parsed, they end up not finding the enum registrations?

Reply:
I think that's how it would go, yes. Although I am not sure how common using thread pools in Dart is. AFAIK the main use case for Dart is with Flutter. Most Flutter apps just run the request async on the UI isolate (thread).
Deferring requests to another isolate requires a so called "send port", which can only transfer simple types between isolates, like a string, but not complex objects. Therefore I don't think its a realistic use case right now.
IMHO we should go with the current implementation and if the need arises, we rebuild the internals of the `EnumRegistry` to use an isolate-agnostic approach.

Reply:
Could we take the opposite and safer approach instead? Use a structure that's thread safe even if it's slower?

Reply:
Sure

Reply:
@baywet Just played around with the package I mentioned above (shared_map).

It only has async methods, which means we would need to make all enum-related methods async too. And for consistency, this would mean all other methods for serializing/deserializing would need to be async as well.

Should I look for another way (I'm not sure there is)? Do we just bite the bullet and make everything async?

Reply:
the alternative to this approach being a serialization function for each enum like we do in other languages (Go, typescript). Is that an option in Dart?

Reply:
Yes. `build_runner` is a common technique to achieve code generation (mostly for JSON). We can either provide builders that the user can call for their project or we delegate the generation to kiota I guess.

Reply:
I see Go uses an `EnumParser` type that takes a string and can return basically anything. It gets passed to the parse nodes and they use it to deserialize enums. Is this function generated by kiota in Go?

Reply:
yes

Reply:
Okay, then these are the next steps:
1. Revert #30 (to make the serialization writer API sync-only again)
2. Change the signature of everything related to enum deser to accept an `EnumParser` (a function that returns an object given a string)
3. Remove the `EnumRegistry`

@baywet can you confirm? Did I get this right?

Reply:
yes, this sounds correct. Thanks for your continuous work on this!

Reply:
@baywet done. This PR should be good then, I think. I basically made everything sync now. The tests show how I imagined the `EnumParser` implementations to look like when they are generated by kiota.

Small aside: the main reason we need to jump through these loops is because we can't access `HttpMethod.values` in a generic way in Dart (see [this](https://github.com/dart-lang/sdk/issues/54136) and [this](https://github.com/dart-lang/language/issues/356)). Hopefully Dart will support abstract static members in the future 🙏

Reply:
I believe the expression is [Jumping through hoops](https://grammarist.com/idiom/jump-through-hoops/), but I couldn't tell for sure, English is not my mother tongue either.

Thanks for making the changes, I believe we're getting close to completion here. Left another comment.